wolverine.py

```diff
@@ -107,47 +122,85 @@ def send_error_to_gpt(file_path, args, error_message, model=DEFAULT_MODEL):
     ]
 
     return json_validated_response(model, messages)
-
-
+    
+# Apply the changes suggested by GPT
 def apply_changes(file_path, changes: list):
     """
     Pass changes as loaded json (list of dicts)
     """
+    try:
+        with open(file_path, "r") as f:
+            original_file_lines = f.readlines()
+
+        operation_changes = [change for change in changes if "operation" in change]
+        explanations = [
+            change["explanation"] for change in changes if "explanation" in change
+        ]
+
+        operation_changes.sort(key=lambda x: x["line"], reverse=True)
+
+        file_lines = original_file_lines.copy()
+        for change in operation_changes:
+            operation = change["operation"]
+            line = change["line"]
+            content = change["content"]
+
+            if operation == "Replace":
+                file_lines[line - 1] = content + "\n"
+            elif operation == "Delete":
+                del file_lines[line - 1]
+            elif operation == "InsertAfter":
+                file_lines.insert(line, content + "\n")
+
+        with open(file_path, "w") as f:
+            f.writelines(file_lines)
+
+        # Print explanations
+        cprint("Explanations:", "blue")
+        for explanation in explanations:
+            cprint(f"- {explanation}", "blue")
+
+        # Show the diff
+        print("\nChanges:")
+        print_diff(original_file_lines, file_lines)
+    except Exception as e:
+        raise Exception(f"Failed to apply changes: {str(e)}")
```

Comment:
what exceptions were you encountering that caused you to add this? Maybe the response wasn't proper JSON (now fixed)?

Reply:
Yeah, I was getting non-real JSON in the output responses, so it wasn't parsing if I recall correctly.